### PR TITLE
interfaces/builtin: add new cuda-support interface

### PR DESCRIPTION
**Pull request by anonymouse64**
While some CUDA specific apparmor code was added to the `opengl` interface in this [PR](https://github.com/snapcore/snapd/pull/5189), there are more permissions that other CUDA programs require, and so I thought that these warranted a new CUDA specific interface. If desired, this could instead be applied to the `opengl` or some other interface, but the devices accessed here are NVIDIA specific so it made the most sense to me to put them in their own interface. Or alternatively, some of the opengl snippets could also be copied into this interface so that a snap that just uses CUDA plugs into `cuda-support` and doesn't need `opengl`.

Example denial for the seccomp:
```
[pid  4349] mknod("/dev/nvidia-uvm-tools", S_IFCHR|0666, makedev(238, 1)) = -1 EPERM (Operation not permitted)
```
Note that the samples seem to run fine without the mknod permission, so if it's too dangerous then it can be removed from the seccomp...

## Testing
To test these, I built a `cuda-samples` snap which bundles almost all of the CUDA samples included with the SDK and allows running them in a test suite. That snap can be found here : https://github.com/anonymouse64/cuda-samples-snap and I would like to integrate that with some automated testing, but I'm not sure how best to go about that as some of the commands create a new x11 window that has to be clicked on or otherwise closed via UI which complicates an automated spread test for example. To test this interface with that snap, after building the snap I do:
```
sudo snap install --jailmode --dangerous cuda-samples*.snap
sudo snap connect cuda-samples:hardware-observe
sudo snap connect cuda-samples:cuda-support
cuda-samples test
# then click on the windows as they come up to close them
```
Some of the sample programs require input files that are copied into the snap, and some of them fail for other reasons that I don't yet understand but I don't think are related to snapd confinement. One such group of samples are the `nvrtc` samples, which seem to fail for reasons not related to confinement (I don't see anything go by in the denials while running but the runtime compilation fails).
Also note you need to have a NVIDIA GPU and appropriate drivers installed to test using the snap. I wasn't able to get noveau drivers to work with CUDA, I had to install the proprietary NVIDIA drivers.

### CODE REVIEW

File: interfaces/builtin/cuda_support.go [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+@{PROC}/sys/vm/mmap_min_addr r,
+@{PROC}/devices r,
+@{PROC}/modules r,
+@{PROC}/driver/nvidia/params r,
+/sys/devices/system/memory/block_size_bytes r,
+unix (bind,listen) type=seqpacket addr="@cuda-uvmfd-[0-9a-f]*",
+/dev/nvidia-uvm wr,
+/dev/nvidiactl wr,
+/{dev,run}/shm/cuda.* rw,
+`
```
Comment (bboozzoo) [inline interfaces/builtin/cuda_support.go]: https://github.com/snapcore/snapd/blob/8378e64cf0931127cad078822c6fe94d983bb659/interfaces/builtin/opengl.go#L39 should be included here and dropped from opengl interface.

File: interfaces/builtin/cuda_support.go [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+const cudaSupportConnectedPlugSecComp = `
+# Description: allow running operations on GPU devices
+# necessary as cuda tries to create /dev/nvidia-uvm-tools with mknod
+mknod - |S_IFCHR
```
Comment (bboozzoo) [inline interfaces/builtin/cuda_support.go]: Isn't this created automatically by udev when nvidia-uvm gets loaded?

Comment (anonymouse64) [reply interfaces/builtin/cuda_support.go]: It is, but CUDA still attempts to create it when running, which is why I think it's okay to just let it try and fail to create this if we don't want to grant this permission to the interface

File: interfaces/builtin/cuda_support.go (hunk 1 above)
Comment (anonymouse64) [reply interfaces/builtin/cuda_support.go]: What about the other nvidida specific libraries? I.e.
```
libnvidia*
tls/libnvidia*
libnvcuvid*
```

**Comment by chipaca** [comment]
What's the status of this work?

**Comment by chipaca** [comment]
I'm adding the "decaying" tag as it's been two months with no activity; I'll close the PR in a week unless you say otherwise.

**Comment by anonymouse64** [comment]
I'm still willing to work on this, still wondering if @bboozzoo can respond to my question about the other nvidia specific library paths.

File: interfaces/builtin/cuda_support.go (hunk 1 above)
Comment (bboozzoo) [reply interfaces/builtin/cuda_support.go]: Sorry, missed the question. I'd say that `libnvidia-tls*` & `tls/libnvidia-tls*` will be needed. The same for `libnvcuvid*`. I'm afraid I do not know if CUDA runtime touches any of the `*GL*` bits. Best would be to start with minimal access to the libraries and run a sample application.

**Comment by jdstrand** [comment]
Thanks for the PR. Many of these are already part of the opengl interface. I suggest adding any others there in a separate PR. For that reason, NAK.

**Comment by jdstrand** [comment]
I'm going to close this PR based on my previous comment. If you feel this is in error, feel free to reopen. Thanks again!

**Comment by anonymouse64** [comment]
@jdstrand The reason I had put this into it's own interface is because of the NVIDIA specific accesses and due to the fact that many CUDA applications, such as machine learning applications, don't need access to all the other paths and such included with the `opengl` interface. If you still think that I should move these accesses into the `opengl` interface I will do that then.

**Comment by chipaca** [comment]
One question (just idle wondering on my part): with my current hardware I could be running a CUDA app while using the intel graphics card for actual graphics. Would this be easier to support if the interfaces are separate? Or does it make no difference?

**Comment by jdstrand** [comment]
I was looking at the accesses only and most of them are included in the opengl interface in recent commits already. I'd prefer to have it all in opengl for now, unless there is a specific access that is meaningfully separated out.

**Comment by anonymouse64** [comment]
@chipaca That would likely be slightly easier to support, if you want you can try playing with my `cuda-samples` snap I made to test these changes with some of the CUDA samples that don't display anything, such as `vectorAdd`, etc. I just tested this and I had to actually add a couple paths from opengl  to `cuda-support` for this to work, specifically this additional apparmor:
```
/dev/nvidia-uvm rw,
/dev/nvidia-uvm-tools rw,
/dev/nvidiactl rw,
/dev/nvidia[0-9]* rw,
/var/lib/snapd/hostfs/{,usr/}lib{,32,64,x32}/{,@{multiarch}/}libnvidia*.so{,.*} rm,
```
I say only slightly, because in these instances you would only have your app plug `cuda-support` instead of the full `opengl` interface. But if your app is using CUDA it will never accidentally try to run on the Intel GPU, if it does I'm pretty sure that's a CUDA bug. And since `opengl` auto-connects, it doesn't seem to make much of a difference to add it to your app.

@jdstrand I guess I mispoke, I meant more than since opengl is auto-connected, there are accesses here in `cuda-support` that we might not necessarily want any program to have such as these for apparmor:

```
/{dev,run}/shm/cuda.* rw,
/dev/nvidia-uvm rw,
/dev/nvidia-uvm-tools rw,
```

and these for seccomp:
```
mknod - |S_IFCHR
```

(though note as mentioned above I'm not really sure if the mknod access is necessary, as the samples all seem to work without that access, probably because as mentioned above loading `nvidia-uvm` should make that unnecessary)

I will also point out I don't know the exact security implications of allowing access to the above AppArmor (and I just noticed the additional `/dev/nvidia-uvm-tools rw` is necessary while re-testing this again so that's missing from this PR diff currently but I can add that easily)

**Pull request closed without merging**